### DISCUSSION TITLE
Milvus 1.1 recovery from WAL error

### DISCUSSION THREAD

**Discussion by vishal-kvn** [milvus-io/milvus#15685] [Q&A and General discussion]
Hi.

I am currently using Milvus 1.1.0 and when I try to start the RO node, I get the following error -
```
Error: Failed to open database: Wal init error!
Milvus server is going to shutdown ...
Milvus server exit...
```

Error Log:
```
tail -f /var/lib/milvus/logs/milvus-22-02-19-03\:37-error.log
[2022-02-19 03:37:50,555][ERROR][WAL][Init][milvus_server] bad wal file 2234
```

By setting `recovery_error_ignore: true`, the RO can be brought up successfully. But I would like to understand if there are any recommended steps to recover the db and not having to set `recovery_error_ignore: true`

Steps followed to recover the db
* Deleting the contents of the `/var/lib/milvus/wal folder did not fix the issue.
* Restarting the node also did not fix the issue.

Once the db gets into this state, the writes are really slow.

Based on previous usage, the db gets into this state when it has concurrent writes. Although I am not able to consistently reproduce the issue following the above steps.

Any help/pointers with this issue will be very helpful.
Thanks in advance.

**Comment by yhmo** [comment]
The error "bad wal file 2234" indicates the file "2234.wal" under the /milvus/wal folder is cracked or missed. The possible reasons:
1. the milvus server unexpectedly terminated or machine power shut down during it is persisting a wal file
2. the file "2234.wal" was deleted from disk before milvus consume it.

Set recovery_error_ignore: true, milvus will ignore the missed file and consume other wal files.

**Comment by yhmo** [comment]
The insert process in server side:
1. receive data from client by RPC
2. write data into wal files
3. copy data into a buffer, wait for flush
4. flush action will persist the data into disk as a segment, and add metadata for the segment
5. after a wal file's data is persisted as segments, the wal file will be deleted from disk

The step 4 is much slower than step 2. If the client continually inserts a lot of data, there will be more and more wal files under the /milvus/wal folder.(We didn't set a limit for the wal files)

**Comment by yhmo** [comment]
When milvus server restarts, it firstly checks the wal folder to see is there any wal files not consumed. It consumes the wal files one by one before providing service.

**Comment by yhmo** [comment]
Since step 4 is slower than step 2. No matter how many clients or threads you use for inserting, the bottleneck is in milvus server, its consumption ability is constant. To keep a reasonable insert speed, you can observe the files count in the wal folder, if the files count is continually increased, that means your client insert too fast.

**Comment by vishal-kvn** [comment]
Thank you for your response. I have a few follow up questions -

- Are WAL files persisted in any other location other than the /wal folder?
- How does Milvus determine the 2234 is missing, is it persisted in an in-memory table/something like etcd?
- Can you please confirm that deleting the contents of the /wal folder and restarting the machine would fix the issue without having to set recovery_error_ignore: true?
- What is the best way to identify the writes or RPC calls that failed due to wal errors?

**Reply by yhmo** [reply]
- All wal files are persisted in /wal folder which defined by the server_config.yaml.
- There is a file "mxlog.meta" under the wal folder, records some flags/states of wal process. When server start, milvus read this file to know which wal file need to be restored.
- Yes, delete the /wal folder can skip this error. But since all the wal files are deleted, maybe there are some data not restored(missed).
- The insert() api return instantly after data is persisted into wal file. If error occurs during persisiting wal file, there will be error logs in the /var/lib/milvus/log/xxx.error.log. Bu the error is not returned by the insert() method.

**Reply by vishal-kvn** [reply]
Thank you for reply. I did notice a behavior after deleting the wal file and restarting the node, new files were created in the /wal folder which made me think that the information regarding wal was being persistent to another location. Have you seen this behavior previously?

**Reply by yhmo** [reply]
After deleting the wal and restart, there should be only two empty files in the wal folder: "lock" and "mxlog.meta".
Maybe some clients continually send insert request to the server?

**Reply by vishal-kvn** [reply]
Thanks for confirmation. The inserts were enabled at that time.

**Reply by nitinrohan19** [reply]
Hi @yhmo, we are now facing a new error while starting multiple RO nodes:
`Loading configuration from: /var/lib/milvus/conf/server_config.yaml
Can't lock file: /var/lib/milvus/wal/lock, due to permission denied. 
Milvus server exit...
single instance lock wal path failed.Can't lock file: /var/lib/milvus/wal/lock, due to permission denied. `

So when we are starting multiple instances of RO node all the instances share the same wal location. And once this lock file is present another RO node accessing it fails to start milvus server. Right now we are not doing any writes so not sure why these files are present in the first place. Also, we tried deleting the lock and mxlog.meta files but as soon as we start the milvus server it again creates these files. I have the following questions:
1. What can we do to avoid this issue or any suggestion would be helpful?
2. Does each RO node have their own separate wal folder or should it be shared across all the RO nodes?
3. How can we completely delete these WAL files permanently so that they don't come up while starting Milvus server

**Reply by yhmo** [reply]
The file "lock" is to avoid multiple milvus processes sharing the same wal folder. Because two servers will write data into a same file and lead to unexpected errors.
The read-only node doesn't write anything into the wal folder, but it also checks the lock file at startup. So different milvus server should use different wal folders.
By the way, milvus server create log file with local time as file name, multiple servers could create same file name with a small chance, so it is better to use different "logs" folder for different servers, too.

Basic rules:
1. multiple milvus servers can share the same "db" folder, but only one write node allowed.
2. different milvus servers should use different "wal" folders
3. different milvus servers should use different "logs" folders
